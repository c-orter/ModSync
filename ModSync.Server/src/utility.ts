import { cpus } from 'node:os';
import path from "node:path";


export class HttpError extends Error {
	constructor(
		public code: number,
		message: string,
	) {
		super(message);
	}

	get codeMessage(): string {
		switch (this.code) {
			case 400:
				return "Bad Request";
			case 404:
				return "Not Found";
			default:
				return "Internal Server Error";
		}
	}
}

export function winPath(p: string): string {
	return p.split(path.posix.sep).join(path.win32.sep);
}

export function unixPath(p: string): string {
	return p.split(path.win32.sep).join(path.posix.sep);
}

/**
 * A lock that is granted when calling [[Semaphore.acquire]].
 */
type Lock = {
	release: () => void
}

/**
 * A task that has been scheduled with a [[Semaphore]] but not yet started.
 */
type WaitingPromise = {
	resolve: (lock: Lock) => void
	reject: (err?: Error) => void
}

/**
 * A [[Semaphore]] is a tool that is used to control concurrent access to a common resource. This implementation
 * is used to apply a max-parallelism threshold.
 */
export class Semaphore {
	private running = 0
	private waiting: WaitingPromise[] = []

	constructor(public max: number = cpus().length) {
		if (max < 1) {
			throw new Error(
				`Semaphore was created with a max value of ${max} but the max value cannot be less than 1`,
			)
		}
	}

	private take() {
		if (this.waiting.length > 0 && this.running < this.max) {
			this.running++

			// Get the next task from the queue
			const task = this.waiting.shift()!

			// Resolve the promise to allow it to start, provide a release function
			task.resolve({ release: this.release })
		}
	}

	public acquire(): Promise<Lock> {
		if (this.running < this.max) {
			this.running++
			return Promise.resolve({ release: this.release })
		}

		return new Promise<Lock>((resolve, reject) => {
			this.waiting.push({ resolve, reject })
		})
	}

	private release = () => {
		this.running--
		this.take()
	}

	/**
	 * Purge all waiting tasks from the [[Semaphore]]
	 */
	public purge() {
		this.waiting.forEach(task => {
			task.reject(
				new Error('The semaphore was purged and as a result this task has been cancelled'),
			)
		})

		this.running = 0
		this.waiting = []
	}
}