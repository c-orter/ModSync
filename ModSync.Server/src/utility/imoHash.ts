import * as fs from "fs";
import * as util from "util";
import { metrohash128 } from "./metroHash";

const SAMPLE_THRESHOLD = 10 * 1024 * 1024;
const SAMPLE_SIZE = 32 * 1024;

// Promisify fs functions
const fsOpen = util.promisify(fs.open);
const fsClose = util.promisify(fs.close);
const fsFstat = util.promisify(fs.fstat);
const fsRead = util.promisify(fs.read);

// Placeholder functions for mmh3 and varint (same as before)

function putUvarint(buf: Uint8Array, x: number): number {
	let i = 0;
	while (x >= 0x80) {
		buf[i] = (x & 0xff) | 0x80;
		x >>= 7;
		i++;
	}
	buf[i] = x & 0xff;
	return i + 1;
}

async function readChunk(
	fd: number,
	position: number,
	length: number,
): Promise<Buffer> {
	const buffer = Buffer.alloc(length);
	const { bytesRead } = await fsRead(fd, buffer, 0, length, position);
	if (bytesRead < length) {
		throw new Error("Could not read enough data");
	}
	return buffer;
}

async function hashFileObject(
	fd: number,
	sampleThreshold: number = SAMPLE_THRESHOLD,
	sampleSize: number = SAMPLE_SIZE,
): Promise<string> {
	const stats = await fsFstat(fd);
	const size = stats.size;

	let data: Buffer;

	if (size < sampleThreshold || sampleSize < 1 || size < 4 * sampleSize) {
		data = await readChunk(fd, 0, size);
	} else {
		const start = await readChunk(fd, 0, sampleSize);
		const middle = await readChunk(fd, Math.floor((size - sampleSize) / 2), sampleSize);
		const end = await readChunk(fd, size - sampleSize, sampleSize);
		data = Buffer.concat([start, middle, end]);
	}

	const hashTmp = metrohash128(data);

	putUvarint(hashTmp, size);

	return Buffer.from(
		hashTmp.buffer,
		hashTmp.byteOffset,
		hashTmp.byteLength,
	).toString("hex");
}

export async function hashFile(
	filename: string,
	sampleThreshold: number = SAMPLE_THRESHOLD,
	sampleSize: number = SAMPLE_SIZE,
): Promise<string> {
	let fd: number | null = null;
	try {
		fd = await fsOpen(filename, "r");
		return await hashFileObject(fd, sampleThreshold, sampleSize);
	} finally {
		if (fd !== null) {
			await fsClose(fd);
		}
	}
}
