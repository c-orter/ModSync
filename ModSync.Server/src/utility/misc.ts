import path from "node:path";

export class HttpError extends Error {
	constructor(
		public code: number,
		message: string,
	) {
		super(message);
	}

	get codeMessage(): string {
		switch (this.code) {
			case 400:
				return "Bad Request";
			case 404:
				return "Not Found";
			default:
				return "Internal Server Error";
		}
	}
}

export function winPath(p: string): string {
	return p.split(path.posix.sep).join(path.win32.sep);
}

export function unixPath(p: string): string {
	return p.split(path.win32.sep).join(path.posix.sep);
}
