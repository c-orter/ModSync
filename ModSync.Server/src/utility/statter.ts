import { stat, type Stats } from "fs-extra";

export interface IStatter {
	stat: (path: string) => Promise<Stats>;
}

export class Statter implements IStatter {
	public stat = stat;
}
