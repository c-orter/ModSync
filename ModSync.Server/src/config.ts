import path from "node:path";
import type { PreSptModLoader } from "@spt/loaders/PreSptModLoader";
import type { JsonUtil } from "@spt/utils/JsonUtil";
import type { VFS } from "@spt/utils/VFS";
import { glob, globNoEnd } from "./glob";
import { unixPath } from "./utility";
import type { ILogger } from "@spt/models/spt/utils/ILogger";

export type SyncPath = {
	path: string;
	enabled?: boolean;
	enforced?: boolean;
	silent?: boolean;
	restartRequired?: boolean;
};

type RawConfig = {
	syncPaths: (string | SyncPath)[];
	commonModExclusions: string[];
};

export class Config {
	constructor(
		public syncPaths: Required<SyncPath>[],
		public commonModExclusions: string[],
	) {}

	public isExcluded(filePath: string): boolean {
		return this.commonModExclusions.some((exclusion) =>
			glob(exclusion).test(unixPath(filePath)),
		);
	}

	public isParentExcluded(filePath: string): boolean {
		return this.commonModExclusions.some((exclusion) =>
			globNoEnd(exclusion).test(unixPath(filePath)),
		);
	}
}
export class ConfigUtil {
	constructor(
		private vfs: VFS,
		private jsonUtil: JsonUtil,
		private modImporter: PreSptModLoader,
		private logger: ILogger,
	) {}

	/**
	 * @throws {Error} If the config file does not exist
	 */
	private readConfigFile(): RawConfig {
		const modPath = this.modImporter.getModPath("Corter-ModSync");
		const configPath = path.join(modPath, "src", "config.jsonc");

		return this.jsonUtil.deserializeJsonC(
			this.vfs.readFile(configPath),
			"config.jsonc",
		) as RawConfig;
	}

	/**
	 * @throws {Error} If the config is invalid
	 */
	private validateConfig(config: RawConfig): void {
		if (!Array.isArray(config.syncPaths))
			throw new Error(
				"Corter-ModSync: config.jsonc 'syncPaths' is not an array. Please verify your config is correct and try again.",
			);

		if (!Array.isArray(config.commonModExclusions))
			throw new Error(
				"Corter-ModSync: config.jsonc 'commonModExclusions' is not an array. Please verify your config is correct and try again.",
			);

		for (const syncPath of config.syncPaths) {
			if (typeof syncPath !== "string" && !("path" in syncPath))
				throw new Error(
					"Corter-ModSync: config.jsonc 'syncPaths' is missing 'path'. Please verify your config is correct and try again.",
				);

			if (
				typeof syncPath === "string"
					? path.isAbsolute(syncPath)
					: path.isAbsolute(syncPath.path)
			)
				throw new Error(
					`Corter-ModSync: SyncPaths must be relative to SPT server root. Invalid path '${syncPath}'`,
				);

			if (
				path
					.relative(
						process.cwd(),
						path.resolve(
							process.cwd(),
							typeof syncPath === "string" ? syncPath : syncPath.path,
						),
					)
					.startsWith("..")
			)
				throw new Error(
					`Corter-ModSync: SyncPaths must within SPT server root. Invalid path '${syncPath}'`,
				);
		}
	}

	public load(): Config {
		const rawConfig = this.readConfigFile();
		this.validateConfig(rawConfig);

		return new Config(
			rawConfig.syncPaths.map((syncPath) => ({
				enabled: true,
				// Not yet implemented
				enforced: false,
				silent: false,
				restartRequired: true,
				...(typeof syncPath === "string" ? { path: syncPath } : syncPath),
			})),
			rawConfig.commonModExclusions,
		);
	}
}
