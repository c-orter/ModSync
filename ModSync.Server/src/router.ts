import type { HttpFileUtil } from "@spt/utils/HttpFileUtil";
import type { SyncUtil } from "./sync";
import { glob } from "./glob";
import type { IncomingMessage, ServerResponse } from "node:http";
import path from "node:path";
import type { VFS } from "@spt/utils/VFS";
import type { Config } from "./config";
import { HttpError, winPath } from "./utility";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import type { PreSptModLoader } from "@spt/loaders/PreSptModLoader";
import type { HttpServerHelper } from "@spt/helpers/HttpServerHelper";

const FALLBACK_SYNCPATHS: Record<string, object> = {
	undefined: ["BepInEx\\plugins\\Corter-ModSync.dll", "ModSync.Updater.exe"],
};

const FALLBACK_HASHES: Record<string, object> = {
	undefined: {
		"BepInEx\\plugins\\Corter-ModSync.dll": { crc: 999999999 },
		"ModSync.Updater.exe": { crc: 999999999 },
	},
};

export class Router {
	constructor(
		private config: Config,
		private syncUtil: SyncUtil,
		private vfs: VFS,
		private httpFileUtil: HttpFileUtil,
		private httpServerHelper: HttpServerHelper,
		private modImporter: PreSptModLoader,
		private logger: ILogger,
	) {}

	/**
	 * @internal
	 */
	public async getServerVersion(
		req: IncomingMessage,
		res: ServerResponse,
		_: RegExpMatchArray,
	) {
		const modPath = this.modImporter.getModPath("Corter-ModSync");
		const packageJson = JSON.parse(
			// @ts-expect-error readFile returns a string when given a valid encoding
			await this.vfs
				// @ts-expect-error readFile takes in an options object, including an encoding option
				.readFilePromisify(path.join(modPath, "package.json"), {
					encoding: "utf-8",
				}),
		);

		res.setHeader("Content-Type", "application/json");
		res.writeHead(200, "OK");
		res.end(JSON.stringify(packageJson.version));
	}

	/**
	 * @internal
	 */
	public async getSyncPaths(
		req: IncomingMessage,
		res: ServerResponse,
		_: RegExpMatchArray,
	) {
		const version = req.headers["modsync-version"] as string;
		if (version in FALLBACK_SYNCPATHS) {
			res.setHeader("Content-Type", "application/json");
			res.writeHead(200, "OK");
			res.end(JSON.stringify(FALLBACK_SYNCPATHS[version]));
			return;
		}

		res.setHeader("Content-Type", "application/json");
		res.writeHead(200, "OK");
		res.end(
			JSON.stringify(
				this.config.syncPaths.map(({ path, ...rest }) => ({
					path: winPath(path),
					...rest,
				})),
			),
		);
	}

	/**
	 * @internal
	 */
	public async getHashes(
		req: IncomingMessage,
		res: ServerResponse,
		_: RegExpMatchArray,
	) {
		const version = req.headers["modsync-version"] as string;
		if (version in FALLBACK_HASHES) {
			res.setHeader("Content-Type", "application/json");
			res.writeHead(200, "OK");
			res.end(JSON.stringify(FALLBACK_HASHES[version]));
			return;
		}

		console.time("hash");
		res.setHeader("Content-Type", "application/json");
		res.writeHead(200, "OK");
		res.end(
			JSON.stringify(await this.syncUtil.hashModFiles(this.config.syncPaths)),
		);
		console.timeEnd("hash");
	}

	/**
	 * @internal
	 */
	public async fetchModFile(
		_: IncomingMessage,
		res: ServerResponse,
		matches: RegExpMatchArray,
	) {
		const filePath = decodeURIComponent(matches[1]);

		const sanitizedPath = this.syncUtil.sanitizeDownloadPath(
			filePath,
			this.config.syncPaths,
		);

		if (!this.vfs.exists(sanitizedPath))
			throw new HttpError(
				404,
				`Attempt to access non-existent path ${filePath}`,
			);

		try {
			const fileStats = await this.vfs.statPromisify(sanitizedPath);
			res.setHeader(
				"Content-Type",
				this.httpServerHelper.getMimeText(path.extname(filePath)) ||
					"text/plain",
			);
			res.setHeader("Content-Length", fileStats.size);
			this.httpFileUtil.sendFile(res, sanitizedPath);
		} catch (e) {
			throw new HttpError(
				500,
				`Corter-ModSync: Error reading '${filePath}'\n${e}`,
			);
		}
	}

	public handleRequest(req: IncomingMessage, res: ServerResponse) {
		const routeTable = [
			{
				route: glob("/modsync/version"),
				handler: this.getServerVersion.bind(this),
			},
			{
				route: glob("/modsync/paths"),
				handler: this.getSyncPaths.bind(this),
			},
			{
				route: glob("/modsync/hashes"),
				handler: this.getHashes.bind(this),
			},
			{
				route: glob("/modsync/fetch/**"),
				handler: this.fetchModFile.bind(this),
			},
		];

		try {
			for (const { route, handler } of routeTable) {
				const matches = route.exec(req.url || "");
				if (matches) return handler(req, res, matches);
			}

			throw new HttpError(404, "Corter-ModSync: Unknown route");
		} catch (e) {
			if (e instanceof Error)
				this.logger.error(
					`Corter-ModSync: Error when handling [${req.method} ${req.url}]:\n${e.message}\n${e.stack}`,
				);

			if (e instanceof HttpError) {
				res.writeHead(e.code, e.codeMessage);
				res.end(e.message);
			} else {
				res.writeHead(500, "Internal server error");
				res.end(
					`Corter-ModSync: Error handling [${req.method} ${req.url}]:\n${e}`,
				);
			}
		}
	}
}
