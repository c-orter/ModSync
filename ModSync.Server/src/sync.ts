import type { VFS } from "@spt/utils/VFS";
import path from "node:path";
import crc32 from "buffer-crc32";
import type { Config, SyncPath } from "./config";
import { HttpError, winPath } from "./utility";
import { readFileSync, statSync } from "node:fs";
import type { ILogger } from "@spt/models/spt/utils/ILogger";

type ModFile = {
	crc: number;
	nosync: boolean;
};

export class SyncUtil {
	constructor(
		private vfs: VFS,
		private config: Config,
		private logger: ILogger,
	) {}

	private getFilesInDir(dir: string): [string, boolean][] {
		if (!this.vfs.exists(dir)) {
			this.logger.warning(
				`Corter-ModSync: Directory '${dir}' does not exist, will be ignored.`,
			);
			return [];
		}
		if (statSync(dir).isFile())
			return [
				[
					dir,
					this.config.isExcluded(dir) ||
						this.vfs.exists(path.join(dir, ".nosync")) ||
						this.vfs.exists(path.join(dir, ".nosync.txt")),
				],
			];

		const nosyncDir =
			this.config.isExcluded(dir) ||
			this.vfs.exists(path.join(dir, ".nosync")) ||
			this.vfs.exists(path.join(dir, ".nosync.txt"));

		return this.vfs
			.getFiles(dir)
			.filter(
				(file) => !file.endsWith(".nosync") && !file.endsWith(".nosync.txt"),
			)
			.map((file): [string, boolean] => [
				path.join(dir, file),
				nosyncDir ||
					this.config.isExcluded(path.join(dir, file)) ||
					this.vfs.exists(`${path.join(dir, file)}.nosync`) ||
					this.vfs.exists(`${path.join(dir, file)}.nosync.txt`),
			])
			.concat(
				this.vfs
					.getDirs(dir)
					.flatMap((subDir) => this.getFilesInDir(path.join(dir, subDir)))
					.map(([child, nosync]): [string, boolean] => [
						child,
						nosyncDir || nosync,
					]),
			);
	}

	private buildModFile(
		file: string,
		// biome-ignore lint/correctness/noEmptyPattern: <explanation>
		{}: Required<SyncPath>,
		nosync: boolean,
	): ModFile {
		try {
			return {
				nosync,
				crc: nosync ? 0 : crc32.unsigned(readFileSync(file)),
			};
		} catch (e) {
			throw new HttpError(500, `Corter-ModSync: Error reading '${file}'\n${e}`);
		}
	}

	public hashModFiles(
		syncPaths: Config["syncPaths"],
	): Record<string, Record<string, ModFile>> {
		return Object.fromEntries(
			syncPaths.map((syncPath) => [
				winPath(syncPath.path),
				Object.fromEntries(
					this.getFilesInDir(syncPath.path).map(
						([file, nosync]) =>
							[
								winPath(file),
								this.buildModFile(file, syncPath, nosync),
							] as const,
					),
				),
			]),
		);
	}

	/**
	 * @throws {Error} If file path is invalid
	 */
	public sanitizeDownloadPath(
		file: string,
		syncPaths: Config["syncPaths"],
	): string {
		const normalized = path.join(
			path.normalize(file).replace(/^(\.\.(\/|\\|$))+/, ""),
		);

		if (
			!syncPaths.some(
				({ path: p }) =>
					!path
						.relative(path.join(process.cwd(), p), normalized)
						.startsWith(".."),
			)
		)
			throw new HttpError(
				400,
				`Corter-ModSync: Requested file '${file}' is not in an enabled sync path!`,
			);

		return normalized;
	}
}
