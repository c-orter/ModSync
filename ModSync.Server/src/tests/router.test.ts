import { expect, beforeEach, describe, it, vi } from "vitest";
import { fs, vol } from "memfs";

import { Router } from "../router";
import { Config } from "../config";
import { SyncUtil } from "../sync";
import { VFS } from "./utils/vfs";
import type { VFS as IVFS } from "@spt/utils/VFS";
import type { HttpFileUtil } from "@spt/utils/HttpFileUtil";
import { mock } from "vitest-mock-extended";
import { PreSptModLoader } from "./utils/preSptModLoader";
import type { PreSptModLoader as IPreSptModLoader } from "@spt/loaders/PreSptModLoader";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import type { ServerResponse } from "node:http";
import { HttpError } from "../utility";

vi.mock("node:fs", () => ({ ...fs, default: fs }));

describe("router", () => {
	const config = new Config(
		[
			{
				path: "plugins",
				enabled: true,
				enforced: false,
				silent: false,
				restartRequired: true,
			},
			{
				path: "user/mods",
				enabled: true,
				enforced: false,
				silent: false,
				restartRequired: false,
			},
			{
				path: "user/cache",
				enabled: false,
				enforced: false,
				silent: false,
				restartRequired: false,
			},
		],
		["plugins/**/node_modules"],
	);
	const vfs = new VFS() as IVFS;
	const logger = mock<ILogger>();
	const syncUtil = new SyncUtil(vfs, config, logger);
	const httpFileUtil = mock<HttpFileUtil>();
	const modImporter = new PreSptModLoader() as IPreSptModLoader;
	const router = new Router(
		config,
		syncUtil,
		vfs,
		httpFileUtil,
		modImporter,
		logger,
	);

	describe("getServerVersion", () => {
		let res: ServerResponse;
		beforeEach(() => {
			vol.reset();
			vol.fromNestedJSON({ "package.json": '{ "version": "1.0.0" }' });

			res = mock<ServerResponse>();
		});

		it("should return server version", () => {
			router.getServerVersion(res, mock<RegExpMatchArray>());

			expect(res.end).toHaveBeenCalledWith(JSON.stringify("1.0.0"));
		});
	});

	describe("getSyncPaths", () => {
		let res: ServerResponse;
		beforeEach(() => {
			res = mock<ServerResponse>();
		});

		it("should return sync paths", () => {
			router.getSyncPaths(res, mock<RegExpMatchArray>());

			expect(res.end).toHaveBeenCalledWith(
				JSON.stringify([
					{
						path: "plugins",
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: true,
					},
					{
						path: "user\\mods",
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: false,
					},
					{
						path: "user\\cache",
						enabled: false,
						enforced: false,
						silent: false,
						restartRequired: false,
					},
				]),
			);
		});
	});

	describe("getHashes", () => {
		const directoryStructure = {
			plugins: {
				"file1.dll": "Test content",
				"file2.dll": "Test content 2",
				"file2.dll.nosync": "",
				"file3.dll": "Test content 3",
				"file3.dll.nosync.txt": "",
				ModName: {
					"mod_name.dll": "Test content 4",
					".nosync": "",
				},
				OtherMod: {
					"other_mod.dll": "Test content 5",
					subdir: {
						"image.png": "Test Image",
						".nosync": "",
					},
				},
			},
			user: {
				mods: {},
			},
		};

		let res: ServerResponse;
		beforeEach(() => {
			vol.reset();
			vol.fromNestedJSON(directoryStructure);

			res = mock<ServerResponse>();
		});

		it("should return hashes", () => {
			router.getHashes(res, mock<RegExpMatchArray>());

			expect(res.end).toHaveBeenCalledWith(
				JSON.stringify({
					plugins: {
						"plugins\\file1.dll": {crc: 1338358949},
						"plugins\\OtherMod\\other_mod.dll": {crc: 2471037616},
					},
					"user\\mods": {},
					"user\\cache": {},
				}),
			);
		});
	});

	describe("fetchModFile", () => {
		const directoryStructure = {
			plugins: {
				"file1.dll": "Test content",
				"file2.dll": "Test content 2",
				"file2.dll.nosync": "",
				"file3.dll": "Test content 3",
				"file3.dll.nosync.txt": "",
				ModName: {
					"mod_name.dll": "Test content 4",
					".nosync": "",
				},
				OtherMod: {
					"other_mod.dll": "Test content 5",
					subdir: {
						"image.png": "Test Image",
						".nosync": "",
					},
				},
			},
			user: {
				mods: {},
			},
		};

		let res: ServerResponse;
		beforeEach(() => {
			vol.reset();
			vol.fromNestedJSON(directoryStructure);
			httpFileUtil.sendFile.mockClear();

			res = mock<ServerResponse>();
		});

		it("should return mod file", () => {
			router.fetchModFile(res, [
				"/modsync/fetch/plugins/file1.dll",
				"plugins/file1.dll",
			]);

			expect(res.setHeader).toHaveBeenCalledWith("Content-Length", 12);
			expect(httpFileUtil.sendFile).toHaveBeenCalledWith(
				res,
				"plugins\\file1.dll",
			);
		});

		it("should reject on non-existent path", () => {
			expect(() => {
				router.fetchModFile(res, [
					"/modsync/fetch/plugins/banana.dll",
					"plugins/banana.dll",
				]);
			}).toThrowError(
				new HttpError(
					404,
					"Attempt to access non-existent path plugins/banana.dll",
				),
			);
		});
	});
});
