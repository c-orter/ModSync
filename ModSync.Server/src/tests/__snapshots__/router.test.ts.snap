// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`router > getHashes > should return hashes 1`] = `
[
  [
    "{"plugins":{"plugins\\\\file1.dll":{"nosync":false,"crc":1338358949},"plugins\\\\file2.dll":{"nosync":true,"crc":0},"plugins\\\\file3.dll":{"nosync":true,"crc":0},"plugins\\\\ModName\\\\mod_name.dll":{"nosync":true,"crc":0},"plugins\\\\OtherMod\\\\other_mod.dll":{"nosync":false,"crc":2471037616},"plugins\\\\OtherMod\\\\subdir\\\\image.png":{"nosync":true,"crc":0}},"user\\\\mods":{},"user\\\\cache":{}}",
  ],
]
`;
